core/sail/fts/lucene-spin/src/main/java/org/eclipse/rdf4j/lucene/spin/LuceneSpinSailConnection.java

```diff
@@ -0,0 +1,261 @@
+		super(wrappedConnection);
+		this.luceneIndex = luceneIndex;
+
+		/*
+		 * Using SailConnectionListener, see <a href="#whySailConnectionListener">above</a>
+		 */
+
+		wrappedConnection.addConnectionListener(connectionListener);
+	}
+
+	@Override
+	public synchronized void addStatement(Resource arg0, IRI arg1, Value arg2, Resource... arg3)
+		throws SailException
+	{
+		super.addStatement(arg0, arg1, arg2, arg3);
+	}
+
+	@Override
+	public void close()
+		throws SailException
+	{
+		if (closed.compareAndSet(false, true)) {
+			try {
+				super.close();
+			}
+			finally {
+				try {
+					luceneIndex.endReading();
+				}
+				catch (IOException e) {
+					logger.warn("could not close IndexReader or IndexSearcher " + e, e);
+				}
+				// remember if you were closed before, some sloppy programmers
+				// may call close() twice.
+			}
+		}
+	}
+
+	// //////////////////////////////// Methods related to indexing
+
+	@Override
+	public synchronized void clear(Resource... arg0)
+		throws SailException
+	{
+		// remove the connection listener, this is safe as the changing methods
+		// are synchronized
+		// during the clear(), no other operation can be invoked
+		getWrappedConnection().removeConnectionListener(connectionListener);
+		try {
+			super.clear(arg0);
+			buffer.clear(arg0);
+		}
+		finally {
+			getWrappedConnection().addConnectionListener(connectionListener);
+		}
+	}
+
+	@Override
+	public void begin()
+		throws SailException
+	{
+		super.begin();
+		buffer.reset();
+		try {
+			luceneIndex.begin();
+		}
+		catch (IOException e) {
+			throw new SailException(e);
+		}
+	}
+
+	@Override
+	public void commit()
+		throws SailException
+	{
+		super.commit();
+
+		logger.debug("Committing Lucene transaction with {} operations.", buffer.operations().size());
+		try {
+			try {
+				// preprocess buffer
+				buffer.optimize();
+
+				// run operations and remove them from buffer
+				for (Iterator<Operation> i = buffer.operations().iterator(); i.hasNext();) {
+					Operation op = i.next();
+					if (op instanceof LuceneSailBuffer.AddRemoveOperation) {
+						AddRemoveOperation addremove = (AddRemoveOperation)op;
+						// add/remove in one call
+						addRemoveStatements(addremove.getAdded(), addremove.getRemoved());
+					}
+					else if (op instanceof LuceneSailBuffer.ClearContextOperation) {
+						// clear context
+						clearContexts(((ClearContextOperation)op).getContexts());
+					}
+					else if (op instanceof LuceneSailBuffer.ClearOperation) {
+						logger.debug("clearing index...");
+						luceneIndex.clear();
+					}
+					else
+						throw new RuntimeException(
+								"Cannot interpret operation " + op + " of type " + op.getClass().getName());
+					i.remove();
+				}
+			}
+			catch (Exception e) {
+				logger.error("Committing operations in lucenesail, encountered exception " + e
+						+ ". Only some operations were stored, " + buffer.operations().size()
+						+ " operations are discarded. Lucene Index is now corrupt.", e);
```

Comment:
Since the index is corrupt, should it be accessible through the Sail anymore? Would it make sense for other operations on the Sail to throw exceptions now rather than return corrupt data?

Also, is there no way to roll back the operations on the lucene index?